File: dart_sample/lib/radio_demo_page.dart
import 'package:flutter/material.dart';

class RadioDemoPage extends StatefulWidget {
  const RadioDemoPage({super.key});

  @override
  State<RadioDemoPage> createState() => _RadioDemoPageState();
}

class _RadioDemoPageState extends State<RadioDemoPage> {
  bool _enabled = true;
  bool _toggleable = true;
  bool _shrinkWrapTapTarget = false;
  String? _materialGroupValue = 'first';
  int _materialChanges = 0;
  String? _adaptiveGroupValue = 'adaptive-first';
  int _adaptiveChanges = 0;
  TargetPlatform _adaptivePlatform = TargetPlatform.iOS;
  bool _adaptiveUseCheckmarkStyle = false;

  @override
  Widget build(BuildContext context) {
    final ThemeData radioTheme = Theme.of(context).copyWith(
      materialTapTargetSize: _shrinkWrapTapTarget
          ? MaterialTapTargetSize.shrinkWrap
          : MaterialTapTargetSize.padded,
    );
    final ThemeData adaptiveTheme = radioTheme.copyWith(
      platform: _adaptivePlatform,
    );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'Radio baseline + adaptive',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Material Radio plus adaptive Cupertino probe with platform/checkmark toggles.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildControlButton(
              label: _enabled ? 'Enabled' : 'Disabled',
              onTap: _toggleEnabled,
              width: 108,
              background: const Color(0xFFE9F0FF),
            ),
            _buildControlButton(
              label: _toggleable ? 'Toggleable' : 'No toggle',
              onTap: _toggleToggleable,
              width: 116,
              background: const Color(0xFFEAE4FF),
            ),
            _buildControlButton(
              label: _shrinkWrapTapTarget ? 'Tap: shrink' : 'Tap: padded',
              onTap: _toggleTapTargetSize,
              width: 128,
              background: const Color(0xFFE8F4E8),
            ),
            _buildControlButton(
              label: 'Reset',
              onTap: _reset,
              width: 80,
              background: const Color(0xFFF3E8D8),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildControlButton(
              label: 'Adaptive: ${_formatPlatform(_adaptivePlatform)}',
              onTap: _cycleAdaptivePlatform,
              width: 156,
              background: const Color(0xFFE7F4FF),
            ),
            _buildControlButton(
              label: _adaptiveUseCheckmarkStyle
                  ? 'Adaptive: checkmark'
                  : 'Adaptive: dot',
              onTap: _toggleAdaptiveCheckmarkStyle,
              width: 164,
              background: const Color(0xFFEFE7FF),
            ),
          ],
        ),
        Text(
          'enabled=$_enabled, toggleable=$_toggleable, materialValue=${_materialGroupValue ?? 'null'}, materialChanges=$_materialChanges, adaptiveValue=${_adaptiveGroupValue ?? 'null'}, adaptiveChanges=$_adaptiveChanges, adaptivePlatform=${_formatPlatform(_adaptivePlatform)}, adaptiveStyle=${_adaptiveUseCheckmarkStyle ? 'checkmark' : 'dot'}, tapTarget=${_shrinkWrapTapTarget ? 'shrinkWrap' : 'padded'}',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Theme(
          data: radioTheme,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 8,
            children: <Widget>[
              const Text(
                'Material path',
                style: TextStyle(fontSize: 13, color: Color(0xFF37474F)),
              ),
              _buildRadioRow(
                radio: Radio<String>(
                  value: 'first',
                  groupValue: _materialGroupValue,
                  onChanged: _enabled ? _onMaterialChanged : null,
                  toggleable: _toggleable,
                ),
                title: 'Default radio #1',
                subtitle: 'value: first',
              ),
              _buildRadioRow(
                radio: Radio<String>(
                  value: 'second',
                  groupValue: _materialGroupValue,
                  onChanged: _enabled ? _onMaterialChanged : null,
                  toggleable: _toggleable,
                ),
                title: 'Default radio #2',
                subtitle: 'value: second',
              ),
              _buildRadioRow(
                radio: Radio<String>(
                  value: 'custom',
                  groupValue: _materialGroupValue,
                  onChanged: _enabled ? _onMaterialChanged : null,
                  toggleable: _toggleable,
                  activeColor: const Color(0xFF00695C),
                  fillColor: WidgetStateProperty.resolveWith((
                    Set<WidgetState> states,
                  ) {
                    if (states.contains(WidgetState.disabled)) {
                      return const Color(0x6100695C);
                    }

                    if (states.contains(WidgetState.selected)) {
                      return const Color(0xFF00695C);
                    }

                    return const Color(0xFF455A64);
                  }),
                  backgroundColor: WidgetStateProperty.resolveWith((
                    Set<WidgetState> states,
                  ) {
                    return states.contains(WidgetState.selected)
                        ? const Color(0x1400695C)
                        : Colors.transparent;
                  }),
                  overlayColor: WidgetStateProperty.resolveWith((
                    Set<WidgetState> states,
                  ) {
                    if (states.contains(WidgetState.pressed)) {
                      return const Color(0x3300695C);
                    }

                    if (states.contains(WidgetState.hovered)) {
                      return const Color(0x2200695C);
                    }

                    if (states.contains(WidgetState.focused)) {
                      return const Color(0x2900695C);
                    }

                    return null;
                  }),
                  side: const BorderSide(color: Color(0xFF00695C), width: 2),
                  innerRadius: WidgetStateProperty.all(5.0),
                ),
                title: 'Custom colors',
                subtitle: 'fill/overlay/side/background overrides',
              ),
              const Text(
                'Adaptive path',
                style: TextStyle(fontSize: 13, color: Color(0xFF37474F)),
              ),
              Theme(
                data: adaptiveTheme,
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  spacing: 8,
                  children: <Widget>[
                    _buildRadioRow(
                      radio: Radio<String>.adaptive(
                        value: 'adaptive-first',
                        groupValue: _adaptiveGroupValue,
                        onChanged: _enabled ? _onAdaptiveChanged : null,
                        toggleable: _toggleable,
                      ),
                      title: 'Adaptive default #1',
                      subtitle: 'value: adaptive-first',
                    ),
                    _buildRadioRow(
                      radio: Radio<String>.adaptive(
                        value: 'adaptive-second',
                        groupValue: _adaptiveGroupValue,
                        onChanged: _enabled ? _onAdaptiveChanged : null,
                        toggleable: _toggleable,
                        activeColor: const Color(0xFF00695C),
                        fillColor: WidgetStateProperty.all(
                          const Color(0xFF8E24AA),
                        ),
                        useCupertinoCheckmarkStyle: _adaptiveUseCheckmarkStyle,
                      ),
                      title: 'Adaptive style probe',
                      subtitle:
                          'checkmark style + fillColor ignore on iOS/macOS',
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildRadioRow({
    required Widget radio,
    required String title,
    required String subtitle,
  }) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      decoration: BoxDecoration(
        color: const Color(0xFFF1F4F9),
        borderRadius: BorderRadius.circular(10),
        border: Border.all(color: const Color(0xFFD6DEEA), width: 1),
      ),
      child: Row(
        spacing: 10,
        children: <Widget>[
          radio,
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              spacing: 2,
              children: <Widget>[
                Text(
                  title,
                  style: const TextStyle(fontSize: 13, color: Colors.black),
                ),
                Text(
                  subtitle,
                  style: const TextStyle(fontSize: 12, color: Colors.black54),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildControlButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onTap,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }

  void _toggleEnabled() {
    setState(() {
      _enabled = !_enabled;
    });
  }

  void _toggleToggleable() {
    setState(() {
      _toggleable = !_toggleable;
    });
  }

  void _toggleTapTargetSize() {
    setState(() {
      _shrinkWrapTapTarget = !_shrinkWrapTapTarget;
    });
  }

  void _cycleAdaptivePlatform() {
    setState(() {
      _adaptivePlatform = switch (_adaptivePlatform) {
        TargetPlatform.iOS => TargetPlatform.macOS,
        TargetPlatform.macOS => TargetPlatform.android,
        _ => TargetPlatform.iOS,
      };
    });
  }

  void _toggleAdaptiveCheckmarkStyle() {
    setState(() {
      _adaptiveUseCheckmarkStyle = !_adaptiveUseCheckmarkStyle;
    });
  }

  void _reset() {
    setState(() {
      _enabled = true;
      _toggleable = true;
      _shrinkWrapTapTarget = false;
      _materialGroupValue = 'first';
      _materialChanges = 0;
      _adaptiveGroupValue = 'adaptive-first';
      _adaptiveChanges = 0;
      _adaptivePlatform = TargetPlatform.iOS;
      _adaptiveUseCheckmarkStyle = false;
    });
  }

  void _onMaterialChanged(String? value) {
    setState(() {
      _materialGroupValue = value;
      _materialChanges += 1;
    });
  }

  void _onAdaptiveChanged(String? value) {
    setState(() {
      _adaptiveGroupValue = value;
      _adaptiveChanges += 1;
    });
  }

  static String _formatPlatform(TargetPlatform platform) {
    return switch (platform) {
      TargetPlatform.iOS => 'iOS',
      TargetPlatform.macOS => 'macOS',
      TargetPlatform.android => 'Android',
      TargetPlatform.fuchsia => 'Fuchsia',
      TargetPlatform.linux => 'Linux',
      TargetPlatform.windows => 'Windows',
    };
  }
}

File: src/Sample/Plumix.Sample/Demos/Cupertino/RadioDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/radio_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class RadioDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new RadioDemoPageState();
    }
}

internal sealed class RadioDemoPageState : State
{
    private bool _enabled = true;
    private bool _toggleable = true;
    private bool _shrinkWrapTapTarget;
    private string? _materialGroupValue = "first";
    private int _materialChanges;
    private string? _adaptiveGroupValue = "adaptive-first";
    private int _adaptiveChanges;
    private TargetPlatform _adaptivePlatform = TargetPlatform.IOS;
    private bool _adaptiveUseCheckmarkStyle;

    public override Widget Build(BuildContext context)
    {
        var baseTheme = Theme.Of(context);
        var radioTheme = baseTheme with
        {
            MaterialTapTargetSize = _shrinkWrapTapTarget
                ? MaterialTapTargetSize.ShrinkWrap
                : MaterialTapTargetSize.Padded
        };
        var adaptiveTheme = radioTheme with
        {
            Platform = _adaptivePlatform
        };

        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("Radio baseline + adaptive", fontSize: 20, color: Colors.Black),
                new Text(
                    "Material Radio plus adaptive Cupertino probe with platform/checkmark toggles.",
                    fontSize: 14,
                    color: Color.Parse("#8A000000")),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildControlButton(
                            label: _enabled ? "Enabled" : "Disabled",
                            onTap: ToggleEnabled,
                            width: 108,
                            background: Color.Parse("#FFE9F0FF")),
                        BuildControlButton(
                            label: _toggleable ? "Toggleable" : "No toggle",
                            onTap: ToggleToggleable,
                            width: 116,
                            background: Color.Parse("#FFEAE4FF")),
                        BuildControlButton(
                            label: _shrinkWrapTapTarget ? "Tap: shrink" : "Tap: padded",
                            onTap: ToggleTapTargetSize,
                            width: 128,
                            background: Color.Parse("#FFE8F4E8")),
                        BuildControlButton(
                            label: "Reset",
                            onTap: Reset,
                            width: 80,
                            background: Color.Parse("#FFF3E8D8")),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildControlButton(
                            label: $"Adaptive: {FormatPlatform(_adaptivePlatform)}",
                            onTap: CycleAdaptivePlatform,
                            width: 156,
                            background: Color.Parse("#FFE7F4FF")),
                        BuildControlButton(
                            label: _adaptiveUseCheckmarkStyle ? "Adaptive: checkmark" : "Adaptive: dot",
                            onTap: ToggleAdaptiveCheckmarkStyle,
                            width: 164,
                            background: Color.Parse("#FFEFE7FF")),
                    ]),
                new Text(
                    $"enabled={(_enabled ? "true" : "false")}, toggleable={(_toggleable ? "true" : "false")}, materialValue={FormatValue(_materialGroupValue)}, materialChanges={_materialChanges}, adaptiveValue={FormatValue(_adaptiveGroupValue)}, adaptiveChanges={_adaptiveChanges}, adaptivePlatform={FormatPlatform(_adaptivePlatform)}, adaptiveStyle={(_adaptiveUseCheckmarkStyle ? "checkmark" : "dot")}, tapTarget={(_shrinkWrapTapTarget ? "shrinkWrap" : "padded")}",
                    fontSize: 12,
                    color: Color.Parse("#FF607D8B")),
                new Theme(
                    data: radioTheme,
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Text(
                                "Material path",
                                fontSize: 13,
                                color: Color.Parse("#FF37474F")),
                            BuildRadioRow(
                                radio: new Radio<string>(
                                    value: "first",
                                    groupValue: _materialGroupValue,
                                    onChanged: _enabled ? OnMaterialChanged : null,
                                    toggleable: _toggleable),
                                title: "Default radio #1",
                                subtitle: "value: first"),
                            BuildRadioRow(
                                radio: new Radio<string>(
                                    value: "second",
                                    groupValue: _materialGroupValue,
                                    onChanged: _enabled ? OnMaterialChanged : null,
                                    toggleable: _toggleable),
                                title: "Default radio #2",
                                subtitle: "value: second"),
                            BuildRadioRow(
                                radio: new Radio<string>(
                                    value: "custom",
                                    groupValue: _materialGroupValue,
                                    onChanged: _enabled ? OnMaterialChanged : null,
                                    toggleable: _toggleable,
                                    activeColor: Color.Parse("#FF00695C"),
                                    fillColor: MaterialStateProperty<Color?>.ResolveWith(states =>
                                    {
                                        if (states.HasFlag(MaterialState.Disabled))
                                        {
                                            return Color.Parse("#6100695C");
                                        }

                                        if (states.HasFlag(MaterialState.Selected))
                                        {
                                            return Color.Parse("#FF00695C");
                                        }

                                        return Color.Parse("#FF455A64");
                                    }),
                                    backgroundColor: MaterialStateProperty<Color?>.ResolveWith(states =>
                                    {
                                        return states.HasFlag(MaterialState.Selected)
                                            ? Color.Parse("#1400695C")
                                            : Colors.Transparent;
                                    }),
                                    overlayColor: MaterialStateProperty<Color?>.ResolveWith(states =>
                                    {
                                        if (states.HasFlag(MaterialState.Pressed))
                                        {
                                            return Color.Parse("#3300695C");
                                        }

                                        if (states.HasFlag(MaterialState.Hovered))
                                        {
                                            return Color.Parse("#2200695C");
                                        }

                                        if (states.HasFlag(MaterialState.Focused))
                                        {
                                            return Color.Parse("#2900695C");
                                        }

                                        return null;
                                    }),
                                    side: new BorderSide(Color.Parse("#FF00695C"), 2),
                                    innerRadius: MaterialStateProperty<double?>.All(5)),
                                title: "Custom colors",
                                subtitle: "fill/overlay/side/background overrides"),
                            new Text(
                                "Adaptive path",
                                fontSize: 13,
                                color: Color.Parse("#FF37474F")),
                            new Theme(
                                data: adaptiveTheme,
                                child: new Column(
                                    crossAxisAlignment: CrossAxisAlignment.Stretch,
                                    spacing: 8,
                                    children:
                                    [
                                        BuildRadioRow(
                                            radio: Radio<string>.Adaptive(
                                                value: "adaptive-first",
                                                groupValue: _adaptiveGroupValue,
                                                onChanged: _enabled ? OnAdaptiveChanged : null,
                                                toggleable: _toggleable),
                                            title: "Adaptive default #1",
                                            subtitle: "value: adaptive-first"),
                                        BuildRadioRow(
                                            radio: Radio<string>.Adaptive(
                                                value: "adaptive-second",
                                                groupValue: _adaptiveGroupValue,
                                                onChanged: _enabled ? OnAdaptiveChanged : null,
                                                toggleable: _toggleable,
                                                activeColor: Color.Parse("#FF00695C"),
                                                fillColor: MaterialStateProperty<Color?>.All(Color.Parse("#FF8E24AA")),
                                                useCupertinoCheckmarkStyle: _adaptiveUseCheckmarkStyle),
                                            title: "Adaptive style probe",
                                            subtitle: "checkmark style + fillColor ignore on iOS/macOS"),
                                    ])),
                        ])),
            ]);
    }

    private Widget BuildRadioRow<T>(Radio<T> radio, string title, string subtitle)
    {
        return new Container(
            padding: new Thickness(10, 8),
            decoration: new BoxDecoration(
                Color: Color.Parse("#FFF1F4F9"),
                BorderRadius: BorderRadius.Circular(10),
                Border: new BorderSide(Color.Parse("#FFD6DEEA"), 1)),
            child: new Row(
                spacing: 10,
                children:
                [
                    radio,
                    new Expanded(
                        child: new Column(
                            crossAxisAlignment: CrossAxisAlignment.Stretch,
                            spacing: 2,
                            children:
                            [
                                new Text(title, fontSize: 13, color: Colors.Black),
                                new Text(subtitle, fontSize: 12, color: Color.Parse("#8A000000")),
                            ])),
                ]));
    }

    private Widget BuildControlButton(
        string label,
        Action onTap,
        double width,
        Color background)
    {
        return new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onTap,
                backgroundColor: background,
                foregroundColor: Colors.Black,
                minHeight: 36,
                padding: new Thickness(10, 8),
                borderRadius: BorderRadius.Circular(8),
                child: new Text(
                    label,
                    fontSize: 12)));
    }

    private void ToggleEnabled()
    {
        SetState(() => _enabled = !_enabled);
    }

    private void ToggleToggleable()
    {
        SetState(() => _toggleable = !_toggleable);
    }

    private void ToggleTapTargetSize()
    {
        SetState(() => _shrinkWrapTapTarget = !_shrinkWrapTapTarget);
    }

    private void CycleAdaptivePlatform()
    {
        SetState(() =>
        {
            _adaptivePlatform = _adaptivePlatform switch
            {
                TargetPlatform.IOS => TargetPlatform.MacOS,
                TargetPlatform.MacOS => TargetPlatform.Android,
                _ => TargetPlatform.IOS,
            };
        });
    }

    private void ToggleAdaptiveCheckmarkStyle()
    {
        SetState(() => _adaptiveUseCheckmarkStyle = !_adaptiveUseCheckmarkStyle);
    }

    private void Reset()
    {
        SetState(() =>
        {
            _enabled = true;
            _toggleable = true;
            _shrinkWrapTapTarget = false;
            _materialGroupValue = "first";
            _materialChanges = 0;
            _adaptiveGroupValue = "adaptive-first";
            _adaptiveChanges = 0;
            _adaptivePlatform = TargetPlatform.IOS;
            _adaptiveUseCheckmarkStyle = false;
        });
    }

    private void OnMaterialChanged(string? value)
    {
        SetState(() =>
        {
            _materialGroupValue = value;
            _materialChanges += 1;
        });
    }

    private void OnAdaptiveChanged(string? value)
    {
        SetState(() =>
        {
            _adaptiveGroupValue = value;
            _adaptiveChanges += 1;
        });
    }

    private static string FormatPlatform(TargetPlatform platform)
    {
        return platform switch
        {
            TargetPlatform.IOS => "iOS",
            TargetPlatform.MacOS => "macOS",
            TargetPlatform.Android => "Android",
            TargetPlatform.Fuchsia => "Fuchsia",
            TargetPlatform.Linux => "Linux",
            TargetPlatform.Windows => "Windows",
            _ => "Unknown",
        };
    }

    private static string FormatValue(string? value)
    {
        return value ?? "null";
    }
}
